### PR TITLE
add solver selection metrics about variable changes

### LINKED ISSUE #13: Measure Squared Change Distance
#12 Use the squared change distance as a metric for solver performance.

### PR DESCRIPTION
**Pull request by j4yk**
Implemented are counting the number of changed variables and, for numbers, the squared change distance. Closes #13 #15

Note that the branch name might be misleading. The branch feature/13_squared_change_distance starts from branch 15_variable_count.

### CODE REVIEW

File: babelsberg/constraintinterpreter.js [hunk 1]
```diff
@@ -269,28 +269,20 @@ Object.subclass('Babelsberg', {
...
-                    minIndex = i;
-                }
-            }
-            if (minIndex > -1) {
-                constraint = constraints[minIndex];
-                console.log('Selected fastest solver:' + constraint.solver.solverName);
-            }
+            constraint = this.chooseConstraintBasedOnMetrics(constraints,
+                    opts.optimizationPriority);
+            console.log('Selected fastest solver:' + constraint.solver.solverName);
```
Comment (psieg) [inline babelsberg/constraintinterpreter.js]: the log has moved to chooseConstraint... and can be removed here

File: babelsberg/constraintinterpreter.js [hunk 2]
```diff
@@ -532,17 +568,23 @@ Object.subclass('Constraint', {
...
             this.constraintvariables.each(function(ea) {
                 var value = ea.getValue();
-                oVariables[ea.ivarname] = value;
+                var oldValue = ea.storedValue;
+                if (oldValue !== value) {
+                    variableAssigments[ea.ivarname] = {oldValue: oldValue,
+                        newValue: value};
+                }
 
+                changedVariables += 1;
```
Comment (psieg) [inline babelsberg/constraintinterpreter.js]: Shouldn't this be inside the if above?

File: babelsberg/constraintinterpreter.js [hunk 3]
```diff
@@ -1099,8 +1154,8 @@ Object.subclass('ConstrainedVariable', {
                     }
             }
 
-            if (this._constraints.length > 0 && !hasEnabledConstraint)
-                return;
+            // if (this._constraints.length > 0 && !hasEnabledConstraint)
+            //     return;
```
Comment (psieg) [inline babelsberg/constraintinterpreter.js]: I don't exactly remember, but wasn't this some magic we added to get tests green and stuff? If the issue is solved now these lines can go entirely

File: babelsberg/tests.js [hunk 4]
```diff
@@ -1922,11 +1922,199 @@ TestCase.subclass('users.timfelgentreff.babelsberg.tests.OnErrorTest', {
...
+		bbb.defaultSolver = this.previousDefaultSolver;
+	},
+});
+
+TestCase.subclass('users.timfelgentreff.babelsberg.tests.AutomaticSolverSelectionDetailsTest', {
+    setUp: function () {
+		this.defaultSolversFixture = new users.timfelgentreff.babelsberg.tests.DefaultSolversFixture();
+		this.defaultSolversFixture.saveDefaultSolvers();
+        bbb.defaultSolvers = [new ClSimplexSolver(), new DBPlanner(), new csp.Solver()];
+        bbb.defaultSolver = null;
```
Comment (psieg) [inline babelsberg/tests.js:1974]: spacing, i'd vote for space in accordance with the primary code

File: babelsberg/tests.js [hunk 5]
```diff
@@ -1922,11 +1922,199 @@ TestCase.subclass('users.timfelgentreff.babelsberg.tests.OnErrorTest', {
...
+                subject: subject
+            },
+            optimizationPriority: ['squaredChangeDistance', 'time'],
+        }, function () {
+            return subject.hat === subject.shoes;
+        });
+        this.assert(subject.hat === subject.shoes);
+    },
+
+});
```
Comment (psieg) [inline babelsberg/tests.js:2195]: The tests do not cover selection through `numberOfChangedVariables`

File: babelsberg/constraintinterpreter.js (hunk 3 above)
Comment (j4yk) [inline babelsberg/constraintinterpreter.js]: Actually, these lines must reappear for the reevaluation, anyway. There is a big BUG comment in a test in #25 to point this out.

File: babelsberg/constraintinterpreter.js [hunk 6]
```diff
@@ -368,6 +359,50 @@ Object.subclass('Babelsberg', {
         return result;
     },
 
+    chooseConstraintBasedOnMetrics: function(constraints, optimizationPriority) {
+        var minIndex = -1;
+        var constraint = null;
+        if (optimizationPriority === undefined) {
+            optimizationPriority = ['time', 'numberOfChangedVariables'];
```
Comment (gersseba) [inline babelsberg/constraintinterpreter.js]: I thought went with "changeDistance" as a name for this metric at some point (as opposed to numeric or squaredChangeDistance for LRueckert's branch) 
It's shorter and more generaly states what is done (what, not how)

File: babelsberg/tests.js (hunk 5 above)
Comment (gersseba) [inline babelsberg/tests.js:2195]: Added tests for numberOfChangedVariables.

**Comment by gersseba** [comment]
@j4yk I added some tests and merged with dev. Do you want to look over it? :+1: from me

**Comment by j4yk** [comment]
Thank you. I would still like to see if the tests stay green with those lines https://github.com/psieg/babelsberg-js/pull/24#discussion-diff-34421726 uncommented again because the other branch needs them anyway. Otherwise, all fine from my point of view.

**Pull request merged**